The app crashes on form finalization

#### ODK Collect version
v2024.3.0-beta.0, current master branch

#### Problem description
https://console.firebase.google.com/u/1/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/7ddd17ca0a93e82d425a5a441cf649de?time=last-seven-days&types=crash&versions=v2024.3.0-beta.0%20(4913)&sessionEventKey=66A3A763019F000148A7574DA36DB8A9_1974385087817821121

```
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.List org.odk.collect.entities.javarosa.parse.EntityFormExtra.getSaveTos()' on a null object reference
	at org.odk.collect.entities.javarosa.finalization.EntityFormFinalizationProcessor.processForm(EntityFormFinalizationProcessor.java:31)
	at org.javarosa.form.api.FormEntryController.lambda$finalizeFormEntry$0$org-javarosa-form-api-FormEntryController(FormEntryController.java:210)
	at org.javarosa.form.api.FormEntryController$$ExternalSyntheticLambda0.accept(Unknown Source:4)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1642)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:606)
	at org.javarosa.form.api.FormEntryController.finalizeFormEntry(FormEntryController.java:209)
	at org.odk.collect.android.javarosawrapper.JavaRosaFormController.finalizeForm(JavaRosaFormController.java:285)
```

#### Steps to reproduce the problem
1. Install Collect v2024.2.1.
2. Download the `All question types` form from the demo server.
3. Open the form and close it.
4. Install v2024.3.0-beta.0 or the current master branch.
5. Open the `All question types` form and finalize it.

#### Expected behavior
There should be no crash.

#### What's going on? 
The problem is that `entityFormExtra` is null after trying to read it from `formDef`
https://github.com/getodk/collect/blob/4741fecf37d167cd97490940a5bd8824546308c6/entities/src/main/java/org/odk/collect/entities/javarosa/finalization/EntityFormFinalizationProcessor.java#L31
Then in the line below we try to use that object and get NPE.

**But how null is returned instead of the expected `entityFormExtra`?**

First of all, we need to keep in mind that we cache `FormDef` after opening a form for the first time. That''s what happened in the third step (see above). The cached `FormDef` contains extras as well.

Ok so now let's understand how we save/read extras. We do that using HashMap where the key is the class name (the fully qualified name of the class):
 
```java
    public void put(T extra) {
        map.put(extra.getClass().getName(), extra);
    }

    public <U extends T> U get(Class<U> key) {
        return (U) map.get(key.getName());
    }
```

So in the third step using v2024.2.1 we save `entityFormExtra` with the key: `org.javarosa.entities.internal.EntityFormExtra`.
However, then we decided that `EntityFormExtra` class along with its whole package should be moved from Javarosa to Collect what was implemented in https://github.com/getodk/collect/pull/6236 and became a part of v2024.3.0-beta.0.

As a result, if we now try to load such extras from cache we try to read them using `org.odk.collect.entities.javarosa.parse.EntityFormExtra` and get null because they exist but are stored with the old class name as a key `org.javarosa.entities.internal.EntityFormExtra`.

--- a/collect_app/src/main/java/org/odk/collect/android/application/initialization/CachedFormsCleaner.kt
+++ b/collect_app/src/main/java/org/odk/collect/android/application/initialization/CachedFormsCleaner.kt
@@ -1,0 +1,23 @@
+package org.odk.collect.android.application.initialization
+
+import org.odk.collect.android.projects.ProjectDependencyModule
+import org.odk.collect.projects.ProjectDependencyFactory
+import org.odk.collect.projects.ProjectsRepository
+import org.odk.collect.upgrade.Upgrade
+import java.io.File
+
+class CachedFormsCleaner(
+    private val projectsRepository: ProjectsRepository,
+    private val projectDependencyModuleFactory: ProjectDependencyFactory<ProjectDependencyModule>
+) : Upgrade {
+    override fun key() = null
+
+    override fun run() {
+        projectsRepository.getAll().forEach { project ->
+            val projectDependencyModule = projectDependencyModuleFactory.create(project.uuid)
+            File(projectDependencyModule.cacheDir)
+                .listFiles { file -> file.name.endsWith(".formdef") }
+                ?.forEach { file -> file.delete() }
+        }
+    }
+}

--- a/collect_app/src/main/java/org/odk/collect/android/application/initialization/upgrade/UpgradeInitializer.kt
+++ b/collect_app/src/main/java/org/odk/collect/android/application/initialization/upgrade/UpgradeInitializer.kt
@@ -2,6 +2,7 @@ package org.odk.collect.android.application.initialization.upgrade
 
 import android.content.Context
 import org.odk.collect.android.BuildConfig
+import org.odk.collect.android.application.initialization.CachedFormsCleaner
 import org.odk.collect.android.application.initialization.ExistingProjectMigrator
 import org.odk.collect.android.application.initialization.ExistingSettingsMigrator
 import org.odk.collect.android.application.initialization.GoogleDriveProjectsDeleter
@@ -18,7 +19,8 @@ class UpgradeInitializer(
     private val existingSettingsMigrator: ExistingSettingsMigrator,
     private val scheduledWorkUpgrade: ScheduledWorkUpgrade,
     private val googleDriveProjectsDeleter: GoogleDriveProjectsDeleter,
-    private val savepointsImporter: SavepointsImporter
+    private val savepointsImporter: SavepointsImporter,
+    private val cachedFormsCleaner: CachedFormsCleaner
 ) {
 
     fun initialize() {
@@ -32,7 +34,8 @@ class UpgradeInitializer(
                 existingSettingsMigrator,
                 scheduledWorkUpgrade,
                 googleDriveProjectsDeleter,
-                savepointsImporter
+                savepointsImporter,
+                cachedFormsCleaner
             )
         ).upgradeIfNeeded()
     }

--- a/collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java
+++ b/collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java
@@ -27,6 +27,7 @@ import org.odk.collect.android.application.CollectSettingsChangeHandler;
 import org.odk.collect.android.application.MapboxClassInstanceCreator;
 import org.odk.collect.android.application.initialization.AnalyticsInitializer;
 import org.odk.collect.android.application.initialization.ApplicationInitializer;
+import org.odk.collect.android.application.initialization.CachedFormsCleaner;
 import org.odk.collect.android.application.initialization.ExistingProjectMigrator;
 import org.odk.collect.android.application.initialization.ExistingSettingsMigrator;
 import org.odk.collect.android.application.initialization.GoogleDriveProjectsDeleter;
@@ -547,7 +548,8 @@ public class AppDependencyModule {
                 existingSettingsMigrator,
                 scheduledWorkUpgrade,
                 googleDriveProjectsDeleter,
-                new SavepointsImporter(projectsRepository, projectDependencyModuleFactory)
+                new SavepointsImporter(projectsRepository, projectDependencyModuleFactory),
+                new CachedFormsCleaner(projectsRepository, projectDependencyModuleFactory)
         );
     }
 

--- a/collect_app/src/test/java/org/odk/collect/android/application/initialization/CachedFormsCleanerTest.kt
+++ b/collect_app/src/test/java/org/odk/collect/android/application/initialization/CachedFormsCleanerTest.kt
@@ -1,0 +1,58 @@
+package org.odk.collect.android.application.initialization
+
+import android.app.Application
+import android.content.Context
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.containsInAnyOrder
+import org.hamcrest.Matchers.equalTo
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.odk.collect.android.injection.DaggerUtils
+import org.odk.collect.projects.Project
+import java.io.File
+
+@RunWith(AndroidJUnit4::class)
+class CachedFormsCleanerTest {
+    private val component = DaggerUtils.getComponent(ApplicationProvider.getApplicationContext<Context>() as Application)
+    private val projectsRepository = component.projectsRepository()
+    private val projectDependencyModuleFactory = component.projectDependencyModuleFactory()
+
+    private val cachedFormsCleaner = CachedFormsCleaner(projectsRepository, projectDependencyModuleFactory)
+
+    @Test
+    fun `cleaner should be run on every app upgrade`() {
+        assertThat(cachedFormsCleaner.key(), equalTo(null))
+    }
+
+    @Test
+    fun `cached forms should be removed on app upgrade from every project`() {
+        projectsRepository.save(Project.Saved("1", "Project 1", "1", "#000000"))
+        projectsRepository.save(Project.Saved("2", "Project 2", "2", "#000000"))
+
+        val project1DependencyModule = projectDependencyModuleFactory.create("1").also {
+            File(it.cacheDir, "file1.formdef").createNewFile()
+            File(it.cacheDir, "file2.formdef").createNewFile()
+            File(it.cacheDir, "file3.save").createNewFile()
+            File(it.cacheDir, "file4.txt").createNewFile()
+        }
+        val project2DependencyModule = projectDependencyModuleFactory.create("2").also {
+            File(it.cacheDir, "file1.formdef").createNewFile()
+            File(it.cacheDir, "file2.formdef").createNewFile()
+            File(it.cacheDir, "file3.save").createNewFile()
+            File(it.cacheDir, "file4.txt").createNewFile()
+        }
+
+        cachedFormsCleaner.run()
+
+        val cachedFilesProject1 = File(project1DependencyModule.cacheDir).listFiles()
+        val cachedFilesProject2 = File(project2DependencyModule.cacheDir).listFiles()
+
+        assertThat(cachedFilesProject1.size, equalTo(2))
+        assertThat(cachedFilesProject1.map { it.name }.toList(), containsInAnyOrder("file3.save", "file4.txt"))
+
+        assertThat(cachedFilesProject2.size, equalTo(2))
+        assertThat(cachedFilesProject2.map { it.name }.toList(), containsInAnyOrder("file3.save", "file4.txt"))
+    }
+}

